src/pip/_internal/resolution/resolvelib/candidates.py

```diff
@@ -197,11 +201,23 @@ def _prepare_abstract_distribution(self):
         # type: () -> AbstractDistribution
         raise NotImplementedError("Override in subclass")
 
+    def _check_metadata_consistency(self):
+        # type: () -> None
+        """Check for consistency of project name and version of dist."""
+        # TODO: (Longer term) Rather than abort, reject this candidate
+        #       and backtrack. This would need resolvelib support.
+        dist = self._dist  # type: Distribution
+        name = canonicalize_name(dist.project_name)
+        if self._name is not None and self._name != name:
+            raise MetadataInconsistent(self._ireq, "name", dist.project_name)
+        version = dist.parsed_version
+        if self._version is not None and self._version != version:
+            raise MetadataInconsistent(self._ireq, "version", dist.version)
```

Comment:
Suggestion for future PRs: keep "moved code" changes separate from functional changes, by making them in separate commits. It makes it easier to keep track of what changed. :)

Reply:
I'll do it for this one too, since I have a rebase to do.